sharray.md

```diff
@@ -0,0 +1,90 @@
+# Sharray
+
+> Sharded IPLD Array
+
+The Sharray is a fixed width IPLD tree structure used to store an array of items.
+
+## Overview
+
+Each node has a height, and a number of items. The number of items must not exceed the trees given width.
+
+The tree must not be sparse, if the tree represents an array of N items, then the left `N/Width` leaves must contain the first `N` items. If `N` is not evenly divisible by `Width` then the final leaf must contain the final remainder. Every node in the tree must have its height set to one less than the node above it. Nodes with a height of 0 contain array values, and nodes with heights greater than zero contain the cids of their child nodes.
```

Comment:
I'm having trouble parsing this. Should this be, "each node must contain (height * width) items"? From the text, it sounds like we need to fit `N` items into `N/Width` slots.

Reply:
Yeah... I had a hard time writing this. I'm basically trying to say every node has at most Width items. And all the items go on the left side of the tree (no gaps). suggestions?

Reply:
The tree must not be sparse. 
Given an array of size `N` and a fixed width of `W`.
- The left `floor(N/W)` leaves contain the first `N` items.
- If `N % W != 0` the final leaf contains the final remainder.
- The tree is perfectly balanced.
- The height is the distance from the leaves, not the root.
- Leaves (nodes with a height of 0) contain array values.
- Inner nodes (nodes with height greater than zero) contain the cids of their child nodes.